ui/src/actions/holdingpen.ts

```diff
@@ -0,0 +1,230 @@
+/* eslint-disable no-underscore-dangle */
+import { push } from 'connected-react-router';
+import { Action, ActionCreator } from 'redux';
+import axios from 'axios';
+
+import { httpErrorToActionPayload } from '../common/utils';
+import {
+  HOLDINGPEN_LOGIN_ERROR,
+  HOLDINGPEN_LOGIN_SUCCESS,
+  HOLDINGPEN_LOGOUT_SUCCESS,
+  HOLDINGPEN_SEARCH_REQUEST,
+  HOLDINGPEN_SEARCH_ERROR,
+  HOLDINGPEN_SEARCH_SUCCESS,
+  HOLDINGPEN_SEARCH_QUERY_UPDATE,
+  HOLDINGPEN_AUTHOR_ERROR,
+  HOLDINGPEN_AUTHOR_REQUEST,
+  HOLDINGPEN_AUTHOR_SUCCESS,
+} from './actionTypes';
+import {
+  BACKOFFICE_API,
+  BACKOFFICE_LOGIN,
+  BACKOFFICE_SEARCH_API,
+  HOLDINGPEN_NEW,
+  HOLDINGPEN_LOGIN_NEW,
+} from '../common/routes';
+import { Credentials } from '../types';
+import storage from '../common/storage';
+import { notifyLoginError } from '../holdingpen-new/notifications';
+import { refreshToken } from '../holdingpen-new/utils/utils';
+
+const httpClient = axios.create();
+
+// Request interceptor for API calls
+httpClient.interceptors.request.use(
+  async (config) => {
+    const token = storage.getSync('holdingpen.token');
+
+    config.headers = {
+      'Content-Type': 'application/json',
+      ...(token && { Authorization: `Bearer ${token}` }),
+    };
+
+    return config;
+  },
+  (error) => Promise.reject(error)
+);
+
+// Response interceptor for API calls
+httpClient.interceptors.response.use(
+  (response) => response,
+  async (error) => {
+    if (axios.isAxiosError(error)) {
+      if (
+        error.response?.status === 403 &&
+        (!(error.config as any)._retry as boolean)
+      ) {
+        (error.config as any)._retry = true;
+        try {
+          const accessToken = await refreshToken();
+          axios.defaults.headers.common.Authorization = `Bearer ${accessToken}`;
+          return httpClient(error.config!);
```

Comment:
```suggestion
          return httpClient(error.config!);
```
It has an extra `!`

Reply:
it's actually a typescript operator (non-null assertion) to remove null and undefined from type (Axios's error.config has a type of ```InternalAxiosRequestConfig<any> | undefined```, and httpClient doesn't accept undefined, thus this operator assures that it can accept it

Reply:
it could be changed to ```return httpClient(error.config || {});```, but then the try catch wouldn't catch ```tokenError``` (that's what I thought when writing this code, but now I'm not sure, please advise 😅)

Reply:
Let's keep it like that with the operator